File: swift/problem541.swift
class Solution {
    func reverseStr(_ s: String, _ k: Int) -> String {
        var s = Array(s)
        let n_max = s.count/(2*k)
        
        if (n_max == 0) {
            s[0..<min(s.count,k)].reverse()
            return String(s)
        }
        
        for n in 0..<n_max {
            s[2*n*k..<(2*n+1)*k].reverse()
        }
        
        if ((s.count%(2*k)) < k) {
            s[(2*n_max*k)...].reverse()
        } else {
            s[(2*n_max*k)..<(2*n_max+1)*k].reverse()
        }
        
        return String(s)
    }
}

File: cpp/problem541.cpp
#include "mystd.h"
using namespace std;

/**
 * why this simple question has so low pass-rate?
 */
class Solution {
public:
    string reverseStr(string s, int k) {
        auto b = s.begin(), e = s.end();
        while( b < e ){
            reverse( b, min(b+k, e) );
            b += k+k;
        }
        return s;
    }
};

int main(int argc, char const *argv[]){
    Solution sol;
    auto s = sol.reverseStr("abcdefg",2);
    cout << s << endl;
    return 0;
}
